cilium/CFP-32759-cilium-mtls-via-mtls-agent.md

```diff
@@ -0,0 +1,260 @@
+
+The socket listening on the mTLS agent has the IP_TRANSPARENT option set, which allows it to bind to a non-local IP address. Note, importantly, that the use of TPROXY rules and IP transparent socket ensures that the destination IP and port are preserved and can be recovered by use of the SO_ORIGINAL_DST option.
+
+![alt text](images/cilium-mTLS-datapath-step-1.png)
+
+#### Step 2: From mTLS Agent A to mTLS Agent B
+
+mTLS Agent A rewrites the destination port of the packet to the well-known port 15008 on the destination mTLS Agent B. The original destination port will be preserved in the request’s header. mTLS Agent A will establish a connection via an [HTTP CONNECT](https://httpwg.org/specs/rfc7540.html#CONNECT) tunnel over mTLS with the mTLS Agent on the same node as the destination pod after establishing a TCP connection with pod A. The packet received is then sent, through the HTTP tunnel, to the host networking stack. From here, the tunneled packet will be routed through node A’s eth0 towards node B. At node B’s eth0, an ingress-to-node BPF program will be invoked, which eventually tail calls the function that implements ingress network policies. All packets with a destination port of 15008 will be considered for redirection.
+
+When such a packet is seen, we check node B’s BPF map for the destination pod’s IP. As Pod B is mTLS-enabled, its IP will be a key in the map. We then look up mTLS Agent B by the well-known key, and again, rewrite the packet’s destination MAC address to that of mTLS Agent B and redirect it to the appropriate network interface.
+
+When the packet reaches mTLS Agent B’s pod network namespace, TPROXY rules in the mTLS agent network namespace will match the packet based on its destination port 15008 and redirect it to 127.0.0.1:15008 via the local loopback device. The packet will reach the process listening on the mTLS Agent at port 15008.
+
+![alt text](images/cilium-mTLS-datapath-step-2.png)
+
+#### Step 3: From mTLS Agent B to Pod B
+
+When the packet reaches mTLS Agent B, we are able to read the request’s header and get the original destination port. If not already established, a new TCP session is established with Pod B at the original destination port. The packet is then routed from mTLS Agent B to Pod B using the standard routing rules of the cluster.
+
+![alt text](images/cilium-mTLS-datapath-step-3.png)
+
+## Requirements
+
+The targeted Cilium release for an MVP of mTLS via transparent proxying is v1.17. 
+
+#### MVP Components
+1. Transparent Redirection
+1. Cilium xDS Server 
+1. mTLS via Transparent Proxy
+
+### MVP Scope
+
+Users can enable Cilium mTLS per workload. All traffic leaving a node from an mTLS-enabled source destined for an mTLS-enabled destination on a different node will be sent over mTLS.
```

Comment:
This is definitely the most common use-case, and certainly the one we want to support _first_, but conceivably, using the policy engine means that we _could_ allow more fine-grained control of the traffic selected for mutual auth.